File: sdks/local-rp/python/tests/test_conformance_dns.py
"""dns.json: `_linkkeys` / `_linkkeys_apis` TXT record parsing, mirroring
`crates/liblinkkeys/src/dns.rs`'s own test cases."""

import pytest

from linkkeys_local_rp import dns as lrp_dns


def test_linkkeys_txt_valid_cases(dns_vectors):
    for case in dns_vectors["linkkeys_txt"]["valid_cases"]:
        record = lrp_dns.parse_linkkeys_txt(case["txt"])
        assert record.fingerprints == case["expected_fingerprints"], case["txt"]


def test_linkkeys_txt_invalid_cases(dns_vectors):
    for case in dns_vectors["linkkeys_txt"]["invalid_cases"]:
        with pytest.raises(lrp_dns.DnsParseError):
            lrp_dns.parse_linkkeys_txt(case["txt"])


def test_linkkeys_txt_no_record_case_is_documentation_only(dns_vectors):
    assert dns_vectors["linkkeys_txt"]["no_record_case"]["documentation_only"] is True


def test_linkkeys_apis_txt_valid_cases(dns_vectors):
    for case in dns_vectors["linkkeys_apis_txt"]["valid_cases"]:
        apis = lrp_dns.parse_linkkeys_apis_txt(case["txt"])
        assert apis.tcp == case["expected_tcp"], case["txt"]
        assert apis.https_base == case["expected_https_base"], case["txt"]


def test_linkkeys_apis_txt_invalid_cases(dns_vectors):
    for case in dns_vectors["linkkeys_apis_txt"]["invalid_cases"]:
        with pytest.raises(lrp_dns.DnsParseError):
            lrp_dns.parse_linkkeys_apis_txt(case["txt"])


def test_default_tcp_port(dns_vectors):
    assert dns_vectors["default_tcp_port"] == lrp_dns.DEFAULT_TCP_PORT

File: sdks/local-rp/ruby/test/test_conformance_dns.rb
# frozen_string_literal: true

require_relative 'test_helper'

class TestConformanceDns < Minitest::Test
  include ConformanceHelper

  def test_default_tcp_port
    assert_equal dns_vector['default_tcp_port'], LinkkeysLocalRp::Dns::DEFAULT_TCP_PORT
  end

  def test_linkkeys_txt_valid_cases
    dns_vector['linkkeys_txt']['valid_cases'].each do |c|
      record = LinkkeysLocalRp::Dns.parse_linkkeys_txt(c['txt'])
      assert_equal c['expected_fingerprints'], record.fingerprints
    end
  end

  def test_linkkeys_txt_invalid_cases
    dns_vector['linkkeys_txt']['invalid_cases'].each do |c|
      assert_raises(LinkkeysLocalRp::Dns::DnsParseError, c['name']) { LinkkeysLocalRp::Dns.parse_linkkeys_txt(c['txt']) }
    end
  end

  def test_linkkeys_apis_txt_valid_cases
    dns_vector['linkkeys_apis_txt']['valid_cases'].each do |c|
      apis = LinkkeysLocalRp::Dns.parse_linkkeys_apis_txt(c['txt'])
      if c['expected_tcp'].nil?
        assert_nil apis.tcp, c['name']
      else
        assert_equal c['expected_tcp'], apis.tcp, c['name']
      end
      if c['expected_https_base'].nil?
        assert_nil apis.https_base, c['name']
      else
        assert_equal c['expected_https_base'], apis.https_base, c['name']
      end
    end
  end

  def test_linkkeys_apis_txt_invalid_cases
    dns_vector['linkkeys_apis_txt']['invalid_cases'].each do |c|
      assert_raises(LinkkeysLocalRp::Dns::DnsParseError, c['name']) do
        LinkkeysLocalRp::Dns.parse_linkkeys_apis_txt(c['txt'])
      end
    end
  end

  def test_no_record_case_is_documentation_only
    # The "no record" case documents DNS-lookup-returns-nothing, not a
    # parser input -- nothing to parse here; the resolver seam itself must
    # treat absence as untrusted (exercised in the flow tests via
    # FakeDnsResolver raising for unknown names).
    assert_nil dns_vector['linkkeys_txt']['no_record_case']['txt']
    assert dns_vector['linkkeys_txt']['no_record_case']['documentation_only']
  end
end
